src/ffi/mod.rs

```diff
@@ -85,12 +119,604 @@ lazy_static::lazy_static! {
+                fwmark: None,
+                name: None,
+                tun: None,
+            })
+            .log_result("Telio::start")
+        })
+    }
+
+    /// Start telio with specified adapter and name.
+    ///
+    /// Adapter will attempt to open its own tunnel.
+    pub fn start_named(
+        &self,
+        private_key: SecretKey,
+        adapter: TelioAdapterType,
+        name: String,
+    ) -> FFIResult<()> {
+        telio_log_info!(
+            "Telio::start entry with instance id: {}. Public key: {:?}. Adapter: {:?}. Name: {}",
+            self.id,
+            private_key.public(),
+            &adapter,
+            &name,
+        );
+        catch_ffi_panic(|| {
+            let mut dev = self.inner.lock().map_err(|_| TelioError::LockError)?;
+            dev.start(&DeviceConfig {
+                private_key,
+                adapter: adapter.into(),
+                fwmark: None,
+                name: Some(name.clone()),
+                tun: None,
+            })
+            .log_result("Telio::start_named")
+        })
+    }
+
+    /// Start telio device with specified adapter and already open tunnel.
+    ///
+    /// Telio will take ownership of tunnel , and close it on stop.
+    ///
+    /// The file descriptor in the `tun` parameter is not forwarded
+    /// on windows
+    ///
+    /// # Parameters
+    /// - `private_key`: base64 encoded private_key.
+    /// - `adapter`: Adapter type.
+    /// - `tun`: A valid filedescriptor to tun device.
+    ///
+    pub fn start_with_tun(
+        &self,
+        private_key: SecretKey,
+        adapter: TelioAdapterType,
+        _tun: i32,
+    ) -> FFIResult<()> {
+        telio_log_info!(
+            "Telio::start entry with instance id: {}. Public key: {:?}. Adapter: {:?}. Tun: {_tun}",
+            self.id,
+            private_key.public(),
+            &adapter
+        );
+        #[cfg(not(target_os = "windows"))]
+        let tun = Some(_tun);
+        #[cfg(target_os = "windows")]
+        let tun = None;
+        catch_ffi_panic(|| {
+            let mut dev = self.inner.lock().map_err(|_| TelioError::LockError)?;
+            dev.start(&DeviceConfig {
+                private_key,
+                adapter: adapter.into(),
+                fwmark: None,
+                name: None,
+                tun,
+            })
+            .log_result("Telio::start_with_tun")
+        })
+    }
+
+    /// Stop telio device.
+    pub fn stop(&self) -> FFIResult<()> {
+        telio_log_info!("Telio::stop entry with instance id: {}.", self.id,);
+        let mut dev = self
+            .inner
+            .lock()
+            .unwrap_or_else(|poisoned| poisoned.into_inner());
+        catch_ffi_panic(|| {
+            dev.stop();
+            Ok(())
+        })
+    }
+
+    /// get device luid.
+    pub fn get_adapter_luid(&self) -> u64 {
+        match self.inner.lock() {
+            Ok(mut d) => d.get_adapter_luid(),
```

Comment:
What will be returned for unix platforms ?

Reply:
`boringtun` and `linux-native` are hard coded to return 0 in `boring.rs` and `linux_native_wg.rs`